cl/opinion_page/utils.py

```diff
@@ -211,6 +232,186 @@ class RelatedCitingResults:
+    ]
+    if settings.RELATED_FILTER_BY_STATUS:
+        # Filter results by status (e.g., Precedential)
+        # Update URL parameters accordingly
+        search_params[
+            f"stat_{PRECEDENTIAL_STATUS.get_status_value(settings.RELATED_FILTER_BY_STATUS)}"
+        ] = True
+        url_search_params = {
+            f"stat_{PRECEDENTIAL_STATUS.get_status_value(settings.RELATED_FILTER_BY_STATUS)}": "on"
+        }
+
+    related_cluster_result = RelatedClusterResults(
+        url_search_params=url_search_params
+    )
+
+    if is_bot(request) or not sub_opinion_pks:
+        return related_cluster_result
+
+    cached_related_clusters, timeout_related = (
+        await cache.aget(mlt_cache_key) or (None, False)
+        if settings.RELATED_USE_CACHE
+        else (None, False)
+    )
+
+    # Prepare related cluster query if not cached results.
+    cluster_search = OpinionClusterDocument.search()
+
+    if cached_related_clusters is not None:
+        related_cluster_result.related_clusters = cached_related_clusters
+        related_cluster_result.timeout = timeout_related
+        related_cluster_result.sub_opinion_pks = list(
+            map(int, sub_opinion_pks)
+        )
+        related_cluster_result.url_search_params = url_search_params
+        return related_cluster_result
+
+    related_query = await build_related_clusters_query(
+        cluster_search, sub_opinion_pks, search_params
+    )
+
+    related_query = related_query.params(
+        timeout=f"{settings.ELASTICSEARCH_FAST_QUERIES_TIMEOUT}s"
+    )
+    related_query = related_query.extra(
+        size=settings.RELATED_COUNT, track_total_hits=False
+    )
+    try:
+        # Execute the Related Query if needed
+        response = related_query.execute()
+        timeout_related = False
+    except (ConnectionError, RequestError, ApiError) as e:
+        logger.warning("Error getting cited and related clusters: %s", e)
+        if settings.DEBUG is True:
+            traceback.print_exc()
+        return related_cluster_result
+    except ConnectionTimeout as e:
+        logger.warning(
+            "ConnectionTimeout getting cited and related clusters: %s", e
+        )
+        response = None
+        timeout_related = True
+
+    related_cluster_result.related_clusters = (
+        response if response is not None else cached_related_clusters or []
+    )
+    related_cluster_result.timeout = False
+    related_cluster_result.sub_opinion_pks = list(map(int, sub_opinion_pks))
+    # related_cluster_result.has_related_cases = True if response else False
+
+    if timeout_related == False:
+        await cache.aset(
+            mlt_cache_key,
+            (related_cluster_result.related_clusters, timeout_related),
+            settings.RELATED_CACHE_TIMEOUT,
+        )
+    return related_cluster_result
+
+
+async def es_get_cited_clusters_with_cache(
+    cluster: OpinionCluster,
+    request: HttpRequest,
+):
```

Comment:
We should add a type hint for the return value of this method. A `dataclass` would be a good choice for this.